### PR TITLE
use sourceless builder for builtin getattr

### PR DESCRIPTION
**Pull request by pmeier**
In TorchVision we use the following (simplified) dispatch mechanism:

```python
import torch


def kernel1(tensor):
    return tensor + 2


def dispatcher1(input):
    kernel = get_kernel(dispatcher1, type(input))
    return kernel(input)


def kernel2(tensor):
    return tensor - 2


def dispatcher2(input):
    kernel = get_kernel(dispatcher2, type(input))
    return kernel(input)


# We actually use the function and type as keys, rather than their names.
# However, this currently not supported, but should be easy to add after
# https://github.com/pytorch/pytorch/pull/111196
REGISTRY = {
    "dispatcher1": {"Tensor": kernel1},
    "dispatcher2": {"Tensor": kernel2},
}


def get_kernel(dispatcher, input_type):
    dispatcher_registry = REGISTRY[dispatcher.__name__]
    for cls in input_type.__mro__:
        kernel = dispatcher_registry[cls.__name__]
        break
    return kernel
```

This can be compiled without graph breaks:

```python
cfn = torch.compile(dispatcher1, fullgraph=True)
torch.testing.assert_close(int(cfn(torch.tensor(3))), 5)

cfn = torch.compile(dispatcher2, fullgraph=True)
torch.testing.assert_close(int(cfn(torch.tensor(3))), 1)
```

However, if we start chaining these calls, we hit some issues:

```python
class Pipeline(torch.nn.Module):
    def forward(self, input):
        input = dispatcher1(input)
        input = dispatcher2(input)
        return input


cfn = torch.compile(Pipeline(), fullgraph=True)
torch.testing.assert_close(int(cfn(torch.tensor(3))), 3)
```

```
Can't access members of type(obj) for a generated custom object. Please use __class__ instead
```

The error message is not really helpful here. The following happens: when compiling `dispatcher1`, `get_kernel` gets inlined. That means when hitting `dispatcher2`, the `type` call no longer happens on an input with a source. Thus, in the first iteration we hit the top branch, while in the second we hit the bottom:

https://github.com/pytorch/pytorch/blob/addb8e29cd842e1a290cb0b55662ee0423ab2498/torch/_dynamo/variables/builtin.py#L1264-L1268

And the error message I posted above originates from the type being treated as constant. This PR replaces this with a `SourcelessBuilder` instead.

With that fix in place, we hit another pointing to `input_type.__mro__`

```
AssertionError: Consider SourcelessBuilder for ephemeral objects, usually objects created locally.
```

Fix is similar: instead of using a `VariableBuilder` here, we use a `SourcelessBuilder` in case we have no `source`:

https://github.com/pytorch/pytorch/blob/addb8e29cd842e1a290cb0b55662ee0423ab2498/torch/_dynamo/variables/builtin.py#L1167-L1168

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4

### CODE REVIEW

File: torch/_dynamo/variables/builtin.py
Comment (peterbell10) [inline torch/_dynamo/variables/builtin.py:1166]: NIT: these cases are handled by the `SourcelessBuilder`
```suggestion
```

**Review by peterbell10 — COMMENTED** [review]
I think this looks reasonable as-is. There are probably some undesirable corner cases like dynamically created classes. However, one of the export test cases show that substituting `type(A)` with `A.__class__` works today and I can't see why they should be treated differently.

File: torch/_dynamo/variables/builtin.py
Comment (pmeier) [reply torch/_dynamo/variables/builtin.py:1166]: Just the three you linked or some more above that you couldn't link because the GH UI wouldn't let you?

File: test/export/test_db.py
Comment (pmeier) [inline test/export/test_db.py:69]: Not sure what to do here. The type reflection is currently the only rewrite case. With that fixed, the parametrization fails, because we don't have anything to parametrize over. However, we probably want to retain the test for the next time someone registers a rewrite case.

One option could be to provide a sentinel value in case the parametrization is empty and fail the test in case we get anything but the sentinel. SOmething like

```python
@parametrize("name,rewrite_case", [...] or [(None, None)])
def test_exportdb_not_supported_rewrite(self, name, rewrite_case):
    assert name is None and rewrite_case is None, "Please remove the default above"
    return
    ...
```

File: torch/_dynamo/variables/builtin.py
Comment (lezcano) [inline torch/_dynamo/variables/builtin.py:1276]: cc @yanboliang Is this alright? The issue torchvision was hitting was taht  `py_type = torch.Tensor` coming from a sourceless tensor, it was wrongly being wrapped into a ConstantVariable rather than a TorchVariable.

Is this alright, or would you want us to be a bit less general here and just do this for `torch.Tensor` so that most paths still hit the UserError?

File: torch/_dynamo/variables/builtin.py
Comment (lezcano) [inline torch/_dynamo/variables/builtin.py:1273]: You can also remove now the TupleVariable if above, as that will be handled by SourcelessBuilder

File: torch/_dynamo/variables/builtin.py
Comment (yanboliang) [reply torch/_dynamo/variables/builtin.py:1276]: ```VariableTrackers``` produced by ```SourcelessBuilder``` are unguarded, what happens if there is a graph break after a ```type(obj)``` call if ```obj``` is generated internally (doesn't have a source)? I think it works well for constant and torch variables, we can't guarantee if they are other variables. But I don't expect a ```type```call returns too complicated data structure, so it should be fine IMO.

**Comment by pmeier** [comment]
Well, @ezyang beat me to it in #113390.

File: torch/_dynamo/variables/builtin.py
Comment (lezcano) [reply torch/_dynamo/variables/builtin.py]: Ok, so Ed argues in https://github.com/pytorch/pytorch/pull/113390/ that this is alright (that PR subsumes this one)

File: torch/_dynamo/variables/builtin.py
Comment (pmeier) [reply torch/_dynamo/variables/builtin.py:1183]: Ok, seems to be just these three lines. At least, I can't find any code for the rest in `SourcelessBuilder.__call__`.

**Review by lezcano — COMMENTED** [review]
Isn't Ed fixing this one?

**Comment by pmeier** [comment]
@lezcano #113390 fixes the `type` issue and parts of `getattr`. However, it does not fix the part of `getattr` that are relevant for the repro above. Plus, there is some cleanup in this PR that is still worth merging.

**Review by peterbell10 — APPROVED** [review]

**Review by lezcano — APPROVED** [review]

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting your change, but the test is failing internally' -c ghfirst

2 test are failing

* `caffe2/test/dynamo:test_dynamo - test_export.py::DynamicShapesExportTests::test_access_class_method_from_user_class_type_fn_<function ExportTests_<lambda> at 0x7fe03d344b80>_dynamic_shapes`
* and `caffe2/test/dynamo:test_dynamo - test_export.py::ExportTests::test_access_class_method_from_user_class_type_fn_<function ExportTests_<lambda> at 0x7fe03d344b80>`

I'm not quite sure what the parameterized lambdas are, but they both fail with no test found error:

```
Test appears to have passed but the binary exited with a non-zero exit code.
This usually means something has crashed after the test was done.
stdout:
============================= test session starts ==============================
platform linux -- Python 3.8.6, pytest-7.2.2, pluggy-0.13.1
rootdir: /data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode
plugins: hypothesis-6.70.1
collected 1171 items

============================ no tests ran in 7.25s =============================

stderr:
Traceback (most recent call last):
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/6081734815403318/caffe2/test/dynamo/__test_dynamo__/test_dynamo#link-tree/__run_lpar_main__.py", line 70, in <module>
    __invoke_main()
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/6081734815403318/caffe2/test/dynamo/__test_dynamo__/test_dynamo#link-tree/__run_lpar_main__.py", line 32, in __invoke_main
    run_as_main(module, main_function)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/6081734815403318/caffe2/test/dynamo/__test_dynamo__/test_dynamo#link-tree/__par__/meta_only/bootstrap.py", line 88, in run_as_main
    oss_run_as_main(
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/6081734815403318/caffe2/test/dynamo/__test_dynamo__/test_dynamo#link-tree/__par__/bootstrap.py", line 58, in run_as_main
    runpy._run_module_as_main(main_module, alter_argv=False)
  File "/usr/local/fbcode/platform010/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/fbcode/platform010/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/6081734815403318/caffe2/test/dynamo/__test_dynamo__/test_dynamo#link-tree/testinfra/testpilot/integration/python/adapters/pytest.py", line 694, in <module>
    sys.exit(main())
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/6081734815403318/caffe2/test/dynamo/__test_dynamo__/test_dynamo#link-tree/testinfra/testpilot/integration/python/adapters/pytest.py", line 690, in main
    return PyTestTestPilotAdapter().run(sys.argv)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/6081734815403318/caffe2/test/dynamo/__test_dynamo__/test_dynamo#link-tree/testinfra/testpilot/integration/python/adapters/base.py", line 322, in run
    return TestPilotCli(test_pilot_adapter=self).run(
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/6081734815403318/caffe2/test/dynamo/__test_dynamo__/test_dynamo#link-tree/testinfra/testpilot/integration/python/adapters/base.py", line 173, in run
    self._execute_tests(test_pilot_test_results, options=options)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/6081734815403318/caffe2/test/dynamo/__test_dynamo__/test_dynamo#link-tree/testinfra/testpilot/integration/python/adapters/base.py", line 132, in _execute_tests
    test_results = self.test_pilot_adapter.execute_tests(
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/6081734815403318/caffe2/test/dynamo/__test_dynamo__/test_dynamo#link-tree/testinfra/testpilot/integration/python/adapters/pytest.py", line 678, in execute_tests
    assert len(results), "No test results returned!"
AssertionError: No test results returned!

exit_code:"Process terminated by the signal: 1"
```

**Comment by pmeier** [comment]
@huydhn Are you sure this is coming from this PR and not #113390? #113390 fixed some behavior with how types are handled by dynamo and in the process adapted the test

https://github.com/pytorch/pytorch/blob/9b0f2f8d94b7485c2c2a4e3d1294a3589b3466b0/test/dynamo/test_export.py#L2854

(the same test name as in your traceback above). Basically there was a check that an error was raised that was removed, since #113390 fixed the underlying issue.

My PR here also touches this test, but only cleans it up and does not change about the way dynamo deals with types.

That being said, if #113390 is reverted as well, we need to wait for it to be relanded before this PR is merged as well.

**Comment by pmeier** [comment]
Or maybe for whatever reason, the internal CI doesn't like lambdas in the parametrization? Not sure how this test came into being, but it seems like a "copy" from the test I linked above?

**Comment by pmeier** [comment]
Here is a diff to avoid having a `lambda` as part of the parametrization

```diff
diff --git a/test/dynamo/test_export.py b/test/dynamo/test_export.py
index 2653f08b4f8..e3cb30cc207 100644
--- a/test/dynamo/test_export.py
+++ b/test/dynamo/test_export.py
@@ -2851,8 +2851,13 @@ def forward(self, x):
         with self.assertRaisesRegex(RuntimeError, "Shape must be more than 4"):
             gm(torch.randn(3, 4, 5))
 
-    @common_utils.parametrize("type_fn", [type, lambda obj: obj.__class__])
-    def test_access_class_method_from_user_class(self, type_fn):
+    @common_utils.parametrize("type_access", ["builtin", "class"])
+    def test_access_class_method_from_user_class(self, type_access):
+        if type_access == "builtin":
+            type_fn = type
+        elif type_access == "class":
+            type_fn = lambda obj: obj.__class__
+
         class A:
             @classmethod
             def func(cls):
```

Not sure why the lambda could be a problem in the parametrization, but maybe worth a try?

**Comment by pmeier** [comment]
Tested my path in https://github.com/pytorch/pytorch/pull/113340#issuecomment-1811993045 with @NicolasHug and it [passed internal checks](https://www.internalfb.com/intern/testinfra/testconsole/testrun/11258999079382887/).

The likely problem was that the lambda in the test name has a memory address in it and if test collection and execution are not happening in the same process, the test is not found.

@pytorchbot merge

**Comment by malfet** [comment]
@pmeier  to be frank, I would like to argue that having lambda address as part of the test name is bad UX
But I thought that `@paramterize` have a mechanism to assign a more repeatable/human understandable names to the test.
Also, I think even test rerun logic we use will not work correctly if test name changes with every materializatoin of the parametrized names

**Comment by lezcano** [comment]
You can use `subtest` to give it a test a static name

**Comment by pmeier** [comment]
@malfet 

> I would like to argue that having lambda address as part of the test name is bad UX
> But I thought that `@paramterize` have a mechanism to assign a more repeatable/human understandable names to the test.

Totally agree here. I made the same assumption as you regarding the test name. But now that I know that this is not the case, I pushed my patch in https://github.com/pytorch/pytorch/pull/113340#issuecomment-1811993045 that gets rid of lambda parametrization, but rather just uses strings.

**Comment by pmeier** [comment]
@lezcano I was mistaken here. I thought you were talking about [`unittest.TestCase.subTest`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.subTest), when you probably actually talking about

https://github.com/pytorch/pytorch/blob/5fb1d8f18ac77f7d5d4a0f0b7a59bd69e0f22d50/torch/testing/_internal/common_utils.py#L381

correct? If so, you are right and I could you used that. Will send a cleanup PR.

**Pull request closed without merging**
